consensus/istanbul/uptime/monitor.go

```diff
@@ -41,51 +34,61 @@ func (u *UptimeEntry) String() string {
 
 // Monitor is responsible for monitoring uptime by processing blocks
 type Monitor struct {
+	epoch          uint64
 	epochSize      uint64
 	lookbackWindow uint64
+	valSetSize     int
+	window         Window
 
-	logger log.Logger
-	store  Store
+	accumulatedUptime *Uptime
+	logger            log.Logger
 }
 
 // NewMonitor creates a new uptime monitor
-func NewMonitor(store Store, epochSize, lookbackWindow uint64) *Monitor {
+func NewMonitor(epochSize, epoch, lookbackWindow uint64, valSetSize int) *Monitor {
+	window := MustMonitoringWindow(epoch, epochSize, lookbackWindow)
+	uptime := new(Uptime)
+	uptime.Entries = make([]UptimeEntry, valSetSize)
+
 	return &Monitor{
-		epochSize:      epochSize,
-		lookbackWindow: lookbackWindow,
-		store:          store,
-		logger:         log.New("module", "uptime-monitor"),
+		epoch:             epoch,
+		epochSize:         epochSize,
+		lookbackWindow:    lookbackWindow,
+		window:            window,
+		accumulatedUptime: uptime,
+		logger:            log.New("module", "uptime-monitor"),
 	}
 }
 
-// MonitoringWindow returns the monitoring window for the given epoch in the format
-// [firstBlock, lastBlock] both inclusive
-func (um *Monitor) MonitoringWindow(epoch uint64) Window {
-	return MustMonitoringWindow(epoch, um.epochSize, um.lookbackWindow)
-}
-
 // ComputeValidatorsUptime retrieves the uptime score for each validator for a given epoch
-func (um *Monitor) ComputeValidatorsUptime(epoch uint64, valSetSize int) ([]*big.Int, error) {
-	logger := um.logger.New("func", "Backend.updateValidatorScores", "epoch", epoch)
+func (um *Monitor) ComputeUptime(header *types.Header) ([]*big.Int, error) {
+	logger := um.logger.New("func", "Backend.updateValidatorScores", "epoch", um.epoch, "until header number", um.accumulatedUptime.LatestHeader.Number.Uint64())
 	logger.Trace("Updating validator scores")
 
-	// The totalMonitoredBlocks are the total number of block on which we monitor uptime for the epoch
-	totalMonitoredBlocks := um.MonitoringWindow(epoch).Size()
+	if istanbul.GetEpochNumber(header.Number.Uint64(), um.epochSize) != um.epoch {
```

Comment:
nit: my though originally is that computing the epoch was "more expensive" than checking if header.number <= last block of epoch (which i can cache). But not entirely sure

Reply:
Yeah, I thought the same, but I changed it also to avoid calling the "isFirstBlock" check. So, although maybe won't improve a lot, I changed it anyways